### PR TITLE
Snapshot credentials so they work on agents

### PR DESCRIPTION
**Pull request by timja**
Fixes https://github.com/jenkinsci/azure-keyvault-plugin/issues/73

Would love to find some time to add some tests around this but we'll see.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/credentials/string/AzureSecretStringCredentials.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public static class DescriptorImpl extends BaseStandardCredentialsDescriptor {
+        @Override
+        @NonNull
+        public String getDisplayName() {
+            return ResourceBundleHolder.get(Messages.class).format("StringCredentialsImpl.secret_text");
+        }
+
+        @Override
+        public boolean isApplicable(CredentialsProvider provider) {
+            return provider instanceof AzureCredentialsProvider;
```
[the remark this answers is not in the document]
Comment (chriskilding) [reply src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/credentials/string/AzureSecretStringCredentials.java:40]: Do you mean that when the user clicks to make a new credential in the UI, they're offered the ability to create an AzureSecretStringCredentials or AWS...StringCredentials? (Which they shouldn't be, because these are read-only types that only the respective providers should be able to instantiate?)

[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/credentials/string/AzureSecretStringCredentials.java:40]: Yes exactly they show up in the UI when creating, the display name is something like `Username With Password` (capital `w` instead of lowercase)

[the remark this answers is not in the document]
Comment (chriskilding) [reply src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/credentials/string/AzureSecretStringCredentials.java:40]: Great find - I've been wondering how to stop that happening for ages 👍

**Comment by jglick** [comment]
Does this actually work? I do not see any new

```java
private Object writeReplace() {
    return Channel.current() == null ? /* XStream */ this : /* Remoting */ CredentialsProvider.snapshot(this);
}
```

Merely implementing `CredentialsSnapshotTaker` accomplishes nothing; in fact AFAIK there is no reason for this API to exist, since you could simply inline your desired snapshot logic directly into `writeReplace`.

**Comment by timja** [comment]
It does work yeah, git plugin works on agent side whereas before it didn’t, this was just copied from the AWS credentials provider

**Comment by jglick** [comment]
I suppose via https://github.com/jenkinsci/git-client-plugin/blob/e57f70993f1b3ce3e319fb498a25d4712de6ec64/src/main/java/org/jenkinsci/plugins/gitclient/RemoteGitImpl.java#L197-L210? Means this will work _only_ for Git credentials, since AFAIK nothing else calls this (there is a usage in CloudBees CI). So for example `withCredentials` will still not work on agents.

**Comment by jglick** [comment]
Correction: `withCredentials` should work with or without this, as secret values would be extracted on the controller: no `Credentials` object is transferred via Remoting in that case. The Git plugins are somewhat unusual in using Remoting overly creatively (which causes problems with `LauncherDecorator` among others), apparently necessitating https://github.com/jenkinsci/git-client-plugin/pull/235. I am not sure offhand what other plugins might be sending `Credentials` instances to an agent. https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#retrieve-the-secret-from-a-specific-credentials-instance and https://github.com/jenkinsci/credentials-plugin/blob/master/docs/implementation.adoc mention it briefly.

(The CloudBees CI call to `CredentialsProvider.snapshot` I mentioned seems to be used to transfer credentials to another controller, which apparently does not require `Credentials.writeReplace` to be overridden, and might not work if it is; hard to follow. @jtnord may know more about the history and usage of this API.)

**Comment by timja** [comment]
Yeah thought I had tested `withCredentials` and it not being an issue.

Do you think it would be better to switch this to `writeReplace`?

**Comment by jglick** [comment]
I guess leave it as is.

**Comment by jtnord** [comment]
Quick comment from my phone will look tomorrow.

Snapshot should be called anywhere before credentials are to be used regardless of agent or controller use.

The credential "secret" may not be loaded at the point in time.  For example in k8s the secret IDs may be known by listing secrets with a specific annotation but we may not load a multi megabyte text field into memory until it is needed.  (Currently I do not do that in the k8s-cred-prov plugin but I should...)

**Comment by jetersen** [comment]
Thanks for the follow up @jtnord I need to implement the same for the hashicorp vault plugin, which is a little complex in that users may already have defined the secrets and I than need to fetch on the controller and snapshot over to the agent 😅
